### PR TITLE
Updated ticket support for /access and /stream endpoints (4-2-stable)

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: irods_specific_implementation/CMakeLists.txt [hunk 1]
```diff
@@ -169,6 +167,7 @@ foreach(API_NAME ${API_SERVER_NAMES})
                         ${IRODS_EXTERNALS_FULLPATH_JSON}/include
                         ${IRODS_EXTERNALS_FULLPATH_BOOST}/include
                         ${IRODS_EXTERNALS_FULLPATH_FMT}/include
+                        /opt/irods-externals/spdlog1.5.0-1/include
```
Comment (korydraughn) [inline irods_specific_implementation/CMakeLists.txt:170]: 4-2-stable does not have spdlog as an externals package.

check if the IRODS_EXTERNALS_PREFIX exists.

File: irods_specific_implementation/irods_rest_access_implementation.h [hunk 2]
```diff
@@ -1,147 +1,168 @@
...
-            access() : api_base{service_name}
-            {
-                // ctor
-            }
+        std::tuple<Pistache::Http::Code &&, std::string>
+        operator()(const Pistache::Http::Header::Collection& _headers,
+                   const std::string& _logical_path,
+                   const std::string& _base_url,
+                   const Pistache::Optional<std::string>& _use_count,
+                   const Pistache::Optional<std::string>& _expires_after)
```
Comment (korydraughn) [inline irods_specific_implementation/irods_rest_access_implementation.h:39]: consider including the word `seconds` in the name. for example:
- _expires_in_seconds
- _expires_after_n_seconds
- _seconds_before_expiry
- **_seconds_until_expiration**
- _good_for_number_of_seconds
- _seconds_until_bad

Comment (korydraughn) [reply irods_specific_implementation/irods_rest_access_implementation.h:39]: use this for the URL parameter name too.

File: irods_specific_implementation/irods_rest_api_base.h [hunk 3]
```diff
@@ -12,231 +13,277 @@
...
+                }
+            }
+
+            return out;
+        } // decode_url
+
+        int set_session_ticket_if_available(const Pistache::Http::Header::Collection& _headers,
+                                            connection_proxy& _conn)
+        {
+            if (const auto h = _headers.tryGetRaw("x-irods-ticket"); !h.isEmpty()) {
```
Comment (korydraughn) [inline irods_specific_implementation/irods_rest_api_base.h:240]: use upper/lower-case consistently.

consider dropping the `x-` prefix.

File: irods_specific_implementation/irods_rest_stream_get_api_implementation.h [hunk 4]
```diff
@@ -1,82 +1,85 @@
...
+        {
+            logger_->trace("Endpoint [{}] initialized.", service_name);
+        }
+
+        std::tuple<Pistache::Http::Code &&, std::string>
+        operator()(const Pistache::Http::Header::Collection& _headers,
+                   const std::string& _body,
+                   const std::string& _path,
+                   const Pistache::Optional<std::string>& _offset,
+                   const Pistache::Optional<std::string>& _limit) // TODO This shouldn't be optional
```
Comment (korydraughn) [inline irods_specific_implementation/irods_rest_stream_get_api_implementation.h:42]: Terrell agrees that this field should not be optional.

File: irods_specific_implementation/irods_rest_access_implementation.h [hunk 5]
```diff
@@ -1,147 +1,168 @@
...
+                           const std::string_view _logical_path) const
+        {
+            rx_ticket(*_conn(), "create", _ticket_id, "read", _logical_path);
+        } // create_ticket
+
+        void set_ticket_use_count(connection_proxy& _conn,
+                                  const std::string_view& _ticket_id,
+                                  const Pistache::Optional<std::string>& _use_count) const
+        {
+            const auto use_count = _use_count.getOrElse("1");
```
Comment (korydraughn) [inline irods_specific_implementation/irods_rest_access_implementation.h:127]: this should default to zero ("0") (i.e. unlimited).

this matches AWS signed URL (canned policy of unlimited use and an expires time).

File: packaging/test_irods_client_rest_cpp.py [hunk 6]
```diff
@@ -45,6 +45,12 @@ def test_access(self):
                 token  = irods_rest.authenticate('rods', 'rods', 'native')
                 result = irods_rest.access(token, logical_path)
                 assert(result.find('error') == -1)
+
+                # Verify that the use count option works as expected.
+                use_count = 50
+                result = json.loads(irods_rest.access(token, logical_path, use_count))
+                ticket_id = result['headers']['X-API-KEY']
```
Comment (korydraughn) [inline packaging/test_irods_client_rest_cpp.py:52]: create an issue about investigating modern header practices around `x-` prefix.

File: irods_specific_implementation/irods_rest_stream_get_api_implementation.h (hunk 4 above)
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_stream_get_api_implementation.h:42]: @trel right now, we parse the limit/bytes_to_read as a 64-bit integer. should we place a cap on the maximum size of the buffer to fill?

File: packaging/test_irods_client_rest_cpp.py (hunk 6 above)
Comment (korydraughn) [reply packaging/test_irods_client_rest_cpp.py:52]: done.

File: irods_specific_implementation/irods_rest_stream_get_api_implementation.h (hunk 4 above)
Comment (trel) [reply irods_specific_implementation/irods_rest_stream_get_api_implementation.h:42]: hmm, an admin configuration option?    with a default that's 'good enough' for known sane use cases...

File: irods_specific_implementation/irods_rest_access_implementation.h (hunk 2 above)
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_access_implementation.h]: done

File: irods_specific_implementation/irods_rest_api_base.h (hunk 3 above)
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_api_base.h]: converted all headers to lowercase and replaced `x-` prefix with `irods-`

File: irods_specific_implementation/irods_rest_access_implementation.h (hunk 5 above)
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_access_implementation.h]: done

File: README.md [hunk 7]
```diff
@@ -288,8 +290,9 @@ Stream data into and out of an iRODS data object
 
 **Parameters**
 - path : The url encoded logical path to a data object
-- offset : The offset in bytes into the data object
-- limit : The maximum number of bytes to read
+- offset : The offset in bytes into the data object. Defaults to 0
+- limit : The maximum number of bytes to read or write. Required for GET
```
Comment (trel) [inline README.md:294]: is there a default here?   i guess not.

Comment (korydraughn) [reply README.md:294]: no default. the cpp rest impl will return an error from the std library.

the current impl is relying on clients to read the API requirements (i.e. the statement saying `limit` is required for GET)

File: README.md [hunk 8]
```diff
@@ -299,11 +302,11 @@ GET : The data requested in the body of the response
 
 **Example CURL Command:**
 ```
-curl -X PUT -H "Authorization: ${TOKEN}" -d"This is some data" "http://localhost/irods-rest/1.0.0/stream?path=%2FtempZone%2Fhome%2Frods%2FfileX&offset=0&limit=1000"
+curl -X PUT -H "Authorization: ${TOKEN}" [-H "X-iRODS-Ticket: ${TICKET}"] -d"This is some data" 'http://localhost/irods-rest/1.0.0/stream?path=%2FtempZone%2Fhome%2Frods%2FfileX&offset=0&limit=1000'
 ```
 or
 ```
-curl -X GET -H "Authorization: ${TOKEN}" "http://localhost/irods-rest/1.0.0/stream?path=%2FtempZone%2Fhome%2Frods%2FfileX&offset=0&limit=1000"
+curl -X GET -H "Authorization: ${TOKEN}" [-H "X-iRODS-Ticket: ${TICKET}"] 'http://localhost/irods-rest/1.0.0/stream?path=%2FtempZone%2Fhome%2Frods%2FfileX&offset=0&limit=1000'
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply README.md]: done

File: README.md [hunk 9]
```diff
@@ -299,11 +302,11 @@ GET : The data requested in the body of the response
 
 **Example CURL Command:**
 ```
-curl -X PUT -H "Authorization: ${TOKEN}" -d"This is some data" "http://localhost/irods-rest/1.0.0/stream?path=%2FtempZone%2Fhome%2Frods%2FfileX&offset=0&limit=1000"
+curl -X PUT -H "Authorization: ${TOKEN}" [-H "X-iRODS-Ticket: ${TICKET}"] -d"This is some data" 'http://localhost/irods-rest/1.0.0/stream?path=%2FtempZone%2Fhome%2Frods%2FfileX&offset=0&limit=1000'
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply README.md]: done.

File: api/StreamPutApi.h [hunk 10]
```diff
@@ -64,7 +62,13 @@ class  StreamPutApi {
     /// <param name="path">irods absolute logical path to the file</param>
     /// <param name="offset">number of bytes to skip for pagination (optional)</param>
     /// <param name="limit">maximum number of bytes to return (optional)</param>
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply api/StreamPutApi.h:64]: done

File: api/StreamGetApi.h [hunk 11]
```diff
@@ -64,7 +64,12 @@ class  StreamGetApi {
     /// <param name="path">irods absolute logical path to the file</param>
     /// <param name="offset">number of bytes to skip for pagination (optional)</param>
     /// <param name="limit">maximum number of bytes to return (optional)</param>
```
Comment (trel) [inline api/StreamGetApi.h:66]: not optional, and switch documentation to match order of call.

File: impl/StreamPutApiImpl.cpp [hunk 12]
```diff
@@ -23,7 +23,13 @@ StreamPutApiImpl::StreamPutApiImpl(Pistache::Address addr)
     : StreamPutApi(addr)
     { }
 
-void StreamPutApiImpl::stream(const Pistache::Http::Header::Collection& headers, const std::string body, const Pistache::Optional<std::string> &path, const Pistache::Optional<std::string> &offset, const Pistache::Optional<std::string> &limit, Pistache::Http::ResponseWriter &response) {
+void StreamPutApiImpl::stream(const Pistache::Http::Header::Collection& headers,
+                              const std::string& body,
+                              const std::string& path,
+                              const Pistache::Optional<std::string>& offset,
```
Comment (trel) [inline impl/StreamPutApiImpl.cpp:29]: now that limit is not optional for streamGet... do we want to keep the same order for streamPut?  aka, limit before offset in the signature?   for consistency?

File: packaging/irods_rest.py [hunk 13]
```diff
@@ -58,7 +58,7 @@ def authenticate(_user_name, _password, _auth_type):
 
     return body
 
-def access(_token, _logical_path):
+def access(_token, _logical_path, _use_count=None, _expires_after=None):
```
Comment (trel) [inline packaging/irods_rest.py:61]: is expires_after still correct here?

File: packaging/irods_rest.py [hunk 14]
```diff
@@ -162,14 +169,18 @@ def put(_token, _physical_path, _logical_path):
...
     offset = 0
     read_size = 1024 * 1024 * 4
     with open(_physical_path, 'w') as f:
         while True:
             c = pycurl.Curl()
             c.setopt(pycurl.HTTPHEADER,['Accept: application/json'])
             c.setopt(pycurl.HTTPHEADER,['Authorization: '+_token])
+
+            if _ticket_id:
+                c.setopt(pycurl.HTTPHEADER,['X-iRODS-Ticket: '+_ticket_id])
```
Comment (trel) [inline packaging/irods_rest.py:182]: irods-ticket?

File: packaging/irods_rest.py [hunk 15]
```diff
@@ -140,6 +143,10 @@ def put(_token, _physical_path, _logical_path):
...
```
Comment (trel) [inline packaging/irods_rest.py:148]: irods-ticket?

File: impl/StreamPutApiImpl.cpp (hunk 12 above)
Comment (korydraughn) [reply impl/StreamPutApiImpl.cpp:29]: yes. that change is already included in the current PR. your comment is just referencing the PUT endpoint.

Comment (trel) [reply impl/StreamPutApiImpl.cpp:29]: yes, wondering if having Put and Get have different parameter orderings will confuse someone.

File: irods_specific_implementation/irods_rest_stream_get_api_implementation.h (hunk 4 above)
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_stream_get_api_implementation.h]: i'm wondering if admins need an option for this at all? do we know if there are clients who allocate 1Gb+ of memory for a single buffer read/write?

wouldn't the size of the buffer depend on the nature of the app and the size of the pipe between cpp rest and the iRODS server?

i'm starting to lean towards using a 32-bit int. with 32 bits, users have a wide range of buffer sizes available and at the same time, they don't run the risk of running out of memory.

here are the posts that have me thinking about this:
- https://stackoverflow.com/a/2811047/5818611
- https://stackoverflow.com/a/41747545/5818611

File: packaging/irods_rest.py (hunk 14 above)
Comment (korydraughn) [reply packaging/irods_rest.py]: done

File: packaging/irods_rest.py (hunk 15 above)
Comment (korydraughn) [reply packaging/irods_rest.py]: done

File: packaging/irods_rest.py (hunk 13 above)
Comment (korydraughn) [reply packaging/irods_rest.py]: done

File: api/StreamGetApi.h (hunk 11 above)
Comment (korydraughn) [reply api/StreamGetApi.h]: done

File: irods_specific_implementation/CMakeLists.txt (hunk 1 above)
Comment (korydraughn) [reply irods_specific_implementation/CMakeLists.txt:170]: we'll need a spdlog externals pkg for 4-2-stable

File: irods_specific_implementation/irods_rest_stream_get_api_implementation.h (hunk 4 above)
Comment (trel) [reply irods_specific_implementation/irods_rest_stream_get_api_implementation.h]: I agree with that assessment.   This is a protection against blowing up the server memory.

Please leave comments to this effect in the defining code.

File: irods_specific_implementation/irods_rest_api_base.h [hunk 16]
```diff
@@ -253,9 +251,9 @@ namespace irods::rest
 
         auto make_error_response(int _error_code, const std::string_view _error_msg) const
         {
-            logger_->error(_error_msg);
-            const auto error = make_error(SYS_INVALID_INPUT_PARAM, _error_msg);
-            return std::forward_as_tuple(Pistache::Http::Code::Bad_Request, error);
+            logger_->error("error_cod={} ::: {}", _error_code, _error_msg);
```
Comment (trel) [inline irods_specific_implementation/irods_rest_api_base.h:254]: error_cod -> error_code

File: irods_specific_implementation/irods_rest_stream_get_api_implementation.h (hunk 4 above)
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_stream_get_api_implementation.h]: done

File: irods_specific_implementation/irods_rest_api_base.h (hunk 16 above)
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_api_base.h]: done

**Comment by korydraughn** [comment]
added #'s

**Comment by korydraughn** [comment]
hold on this, I need to change the spdlog cmake reference to the cmakelists file

**Comment by trel** [comment]
if we ship this as is - we can ship against 4.2.9

**Comment by trel** [comment]
then we can bump and ship against 4.2.11 with the CMAKE variable

**Comment by trel** [comment]
confirmed - merging.

**Pull request merged**